src/connectors/impls/gpubsub/gsub.rs

```diff
@@ -0,0 +1,331 @@
+// Copyright 2022, The Tremor Team
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use crate::connectors::google::AuthInterceptor;
+use crate::connectors::prelude::*;
+use crate::connectors::utils::url::HttpsDefaults;
+use async_std::channel::{Receiver, Sender};
+use async_std::stream::StreamExt;
+use async_std::task::JoinHandle;
+use dashmap::DashMap;
+use googapis::google::pubsub::v1::subscriber_client::SubscriberClient;
+use googapis::google::pubsub::v1::{PubsubMessage, ReceivedMessage, StreamingPullRequest};
+use gouth::Token;
+use serde::Deserialize;
+use std::fmt::Debug;
+use std::sync::Arc;
+use std::time::Duration;
+use tonic::codegen::InterceptedService;
+use tonic::transport::{Certificate, Channel, ClientTlsConfig};
+use tonic::Status;
+use tremor_pipeline::ConfigImpl;
+
+#[derive(Deserialize, Clone)]
+struct Config {
+    pub connect_timeout: u64,
+    pub ack_deadline: u64,
+    pub subscription_id: String,
+    #[serde(default = "default_endpoint")]
+    pub endpoint: String,
+    #[serde(default = "default_skip_authentication")]
+    pub skip_authentication: bool,
+}
+impl ConfigImpl for Config {}
+
+fn default_endpoint() -> String {
+    "https://pubsub.googleapis.com".into()
+}
+
+fn default_skip_authentication() -> bool {
+    false
+}
+
+#[derive(Debug, Default)]
+pub(crate) struct Builder {}
+
+#[async_trait::async_trait]
+impl ConnectorBuilder for Builder {
+    fn connector_type(&self) -> ConnectorType {
+        "gsub".into()
+    }
+
+    async fn build(&self, alias: &str, raw_config: &ConnectorConfig) -> Result<Box<dyn Connector>> {
+        if let Some(raw) = &raw_config.config {
+            let config = Config::new(raw)?;
+            Ok(Box::new(GSub { config }))
+        } else {
+            Err(ErrorKind::MissingConfiguration(alias.to_string()).into())
+        }
+    }
+}
+
+struct GSub {
+    config: Config,
+}
+
+type PubSubClient = SubscriberClient<InterceptedService<Channel, AuthInterceptor>>;
+type AsyncTaskMessage = Result<(u64, Vec<u8>)>;
+
+struct GSubSource {
+    config: Config,
+    client: Option<PubSubClient>,
+    receiver: Option<Receiver<AsyncTaskMessage>>,
+    ack_sender: Option<Sender<u64>>,
+    ack_ids: Arc<DashMap<u64, String>>,
```

Comment:
i think you don't need this anymore here, do you? You could change the map in the `consumer_task` to be a simple hashmap, as it is contained in that task. Or would this make us lose `ack_ids` in case of a reconnect?

Reply:
yep, we would lose the ACK IDs, OTOH it's not necessairly a big issue, since then we'd just process the messages again, and as reconnects are rather rare, I don't think this would be an issue (we don't guarantee exactly once delivery anyway, as far as I know)